11_FFTBloom/app_resources/fft_convolve_ifft.hlsl

```diff
@@ -0,0 +1,204 @@
+				complex_t<scalar_t> nyquist = getDFTMirror<SharedmemAdaptor>(localElementIndex, sharedmemAdaptor);
+
+				workgroup::fft::unpack<scalar_t>(zero, nyquist);
+
+				// We now have zero and Nyquist frequencies at NFFT[index], so we must use `getDFTIndex(index)` to get the actual index into the DFT
+				const uint32_t globalElementIndex = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | workgroup::SubgroupContiguousIndex();
+				const uint32_t indexDFT = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getDFTIndex(globalElementIndex);
+
+				float32_t2 uv = float32_t2(indexDFT / float32_t(FFT_LENGTH), float32_t(0)) + pushConstants.kernelHalfPixelSize;
+				const vector<scalar_t, 2> zeroKernelVector = kernelChannels[channel].SampleLevel(samplerState[channel], uv, 0);
+				const complex_t<scalar_t> zeroKernel = { zeroKernelVector.x, zeroKernelVector.y };
+				zero = zero * zeroKernel;
+
+				// Do the same for the nyquist coord
+				uv.y += 0.5;
+				const vector<scalar_t, 2> nyquistKernelVector = kernelChannels[channel].SampleLevel(samplerState[channel], uv, 0);
+				const complex_t<scalar_t> nyquistKernel = { nyquistKernelVector.x, nyquistKernelVector.y };
+				nyquist = nyquist * nyquistKernel;
+
+				// Since their IFFT is going to be real, we can pack them back as Z + iN, do a single IFFT and recover them afterwards
+				preloaded[localElementIndex] = zero + rotateLeft<scalar_t>(nyquist);
+
+				// We have set Z + iN for an even element (lower half of the DFT). We must now set conj(Z) + i * conj(N) for an odd element (upper half of DFT)
+				// The logic here is basically the same as in getDFTMirror: we figure out which of our odd elements corresponds to the other thread's
+				// current even element (current even element is `localElementIndex` and our local odd element that's the mirror of the other thread's even element is
+				// `elementToTradeLocalIdx`. Then we get conj(Z) + i * conj(N) from that thread and send our own via a shuffle
+				const complex_t<scalar_t> mirrored = conj(zero) + rotateLeft<scalar_t>(conj(nyquist));
+				vector<scalar_t, 2> mirroredVector = { mirrored.real(), mirrored.imag() };
+				const uint32_t otherElementIdx = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getNablaMirrorIndex(globalElementIndex);
+				const uint32_t otherThreadID = otherElementIdx & (_NBL_HLSL_WORKGROUP_SIZE_ - 1);
+				const uint32_t otherThreadGlobalElementIndex = _NBL_HLSL_WORKGROUP_SIZE_ * localElementIndex | otherThreadID;
+				const uint32_t elementToTradeGlobalIdx = workgroup::fft::FFTIndexingUtils<ELEMENTS_PER_THREAD, _NBL_HLSL_WORKGROUP_SIZE_>::getNablaMirrorIndex(otherThreadGlobalElementIndex);
+				const uint32_t elementToTradeLocalIdx = elementToTradeGlobalIdx / _NBL_HLSL_WORKGROUP_SIZE_;
+				workgroup::Shuffle<SharedmemAdaptor, vector<scalar_t, 2> >::__call(mirroredVector, otherThreadID, sharedmemAdaptor);
```

Comment:
1. why isn't this rolled up into some function or functor in the `workgroup` namespace?

call it `pack` or something, do you really think that the user should know/remember how to write such a complex thing?

Reply:
This goes inside of the preloaded accessor rn, but I could make a generic version for a function that expects index, adaptor AND the element instead